packages/token-search-discovery-controller/src/token-search-discovery-controller.ts

```diff
@@ -0,0 +1,122 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { TokenSearchResponseItem } from './types';
+
+export const TOKENSEARCH_EVENTS = {
+  SEARCH_COMPLETED: 'SEARCH_COMPLETED',
+} as const;
+
+const name = 'TokenSearchDiscoveryController';
+
+type Events = {
+  type: (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS];
+  payload: [TokenSearchResponseItem[]];
+};
+
+export type TokenSearchDiscoveryControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof name,
+    never,
+    Events,
+    never,
+    (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS]
+  >;
+
+export type TokenSearchDiscoveryState = {
+  recentSearches: TokenSearchResponseItem[];
+  lastSearchTimestamp: number | null;
+};
+
+const defaultState: TokenSearchDiscoveryState = {
+  recentSearches: [],
+  lastSearchTimestamp: null,
+};
+
+export class TokenSearchDiscoveryController extends BaseController<
+  typeof name,
+  TokenSearchDiscoveryState,
+  TokenSearchDiscoveryControllerMessenger
+> {
+  private readonly baseUrl: string;
+
+  constructor({
+    portfolioApiUrl,
+    initialState,
+    messenger,
+  }: {
+    portfolioApiUrl: string;
+    initialState?: TokenSearchDiscoveryState;
+    messenger: TokenSearchDiscoveryControllerMessenger;
+  }) {
+    super({
+      name,
+      metadata: {
+        recentSearches: { persist: true, anonymous: false },
+        lastSearchTimestamp: { persist: true, anonymous: false },
+      },
+      messenger,
+      state: { ...defaultState, ...initialState },
+    });
+
+    if (!portfolioApiUrl) {
+      throw new Error('Portfolio API URL is not set');
+    }
+    this.baseUrl = portfolioApiUrl;
+  }
+
+  async searchTokens(
+    chains: string[] = [],
+    name?: string,
+    limit?: string,
```

Comment:
Since you have three optional arguments, what happens if the consumer wants to pass, say, the third argument but not the second? Then they'd have to pass `undefined`. That might be awkward, so what are your thoughts on placing all of these arguments in an (optional) options bag? For example:

``` suggestion
    {
      chains = [],
      name,
      limit,
    }: {
      chains?: string[],
      name?: string,
      limit?: string,
    }